d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterPropertiesJsonSerializer.java

```diff
@@ -157,6 +168,21 @@ public ClusterProperties fromMap(Map<String, Object> map)
       delegated = mapGet(map, PropertyKeys.DELEGATED);
     }
 
+    // get canary cluster properties and canary distribution strategy, if exist
+    Map<String, Object> canaryConfigsMap = mapGetOrDefault(map, PropertyKeys.CANARY_CONFIGS, Collections.emptyMap());
+    Map<String, Object> distributionStrategyMap = mapGetOrDefault(map, PropertyKeys.CANARY_DISTRIBUTION_STRATEGY, Collections.emptyMap());
+    if (!canaryConfigsMap.isEmpty() && !distributionStrategyMap.isEmpty()) {
```

Comment:
How about just checking for the existence of keys in the `map` instead of creating `emptyMap` objects?

Reply:
I think the difference between the current way and checking keys for existence is just to move lines 172 & 173 around. You will have:

if (map.containsKey(KEY1) && map.containsKey(KEY2) {
   Map<String, Object> canaryConfigsMap = mapGet(map, PropertyKeys.CANARY_CONFIGS);
   Map<String, Object> distributionStrategyMap = mapGet(map, PropertyKeys.CANARY_DISTRIBUTION_STRATEGY);
   ... same code next ....
}

, which is not much different but just a personal reference?

Reply:
I think the difference is when the map doesn't contain the keys, your code will allocate empty maps at run time, which may be less efficient. This is a minor issue anyways. Up to you to make the final call

```
if (map.containsKey(KEY1) && map.containsKey(KEY2) {
Map<String, Object> canaryConfigsMap = mapGet(map, PropertyKeys.CANARY_CONFIGS);
Map<String, Object> distributionStrategyMap = mapGet(map, PropertyKeys.CANARY_DISTRIBUTION_STRATEGY);
... same code next ....
}
```

This looks pretty clear to me